tripy/tests/spec_verification/test_dtype_constraints.py

```diff
@@ -17,118 +17,188 @@
-    """
-    handles any function that has a special way of being called such as methods
-    """
+def _method_handler(func_name, kwargs, func_obj, api_call_locals):
     _METHOD_OPS = {
-        "__add__": (lambda a, b: f"{a} + {b}"),
-        "__sub__": (lambda a, b: f"{a} - {b}"),
-        "__rsub__": (lambda a, b: f"{a} - {b}"),
-        "__pow__": (lambda a, b: f"{a}**{b}"),
-        "__rpow__": (lambda a, b: f"{a}**{b}"),
-        "__mul__": (lambda a, b: f"{a} * {b}"),
-        "__rmul__": (lambda a, b: f"{a} * {b}"),
-        "__truediv__": (lambda a, b: f"{a} / {b}"),
-        "__lt__": (lambda a, b: f"{a} < {b}"),
-        "__le__": (lambda a, b: f"{a} <= {b}"),
-        "__eq__": (lambda a, b: f"{a} == {b}"),
-        "__ne__": (lambda a, b: f"{a} != {b}"),
-        "__ge__": (lambda a, b: f"{a} >= {b}"),
-        "__gt__": (lambda a, b: f"{a} > {b}"),
+        "__add__": (lambda self, other: self + other),
+        "__sub__": (lambda self, other: self - other),
+        "__rsub__": (lambda self, other: self - other),
+        "__pow__": (lambda self, other: self**other),
+        "__rpow__": (lambda self, other: self**other),
+        "__mul__": (lambda self, other: self * other),
+        "__rmul__": (lambda self, other: self * other),
+        "__truediv__": (lambda self, other: self / other),
+        "__rtruediv__": (lambda self, other: self / other),
+        "__lt__": (lambda self, other: self < other),
+        "__le__": (lambda self, other: self <= other),
+        "__eq__": (lambda self, other: self == other),
+        "__ne__": (lambda self, other: self != other),
+        "__ge__": (lambda self, other: self >= other),
+        "__gt__": (lambda self, other: self > other),
+        "__matmul__": (lambda self, other: self @ other),
+        "shape": (lambda self: self.shape),
+        "__getitem__": (lambda self, index: self.__getitem__(index)),
     }
-    if not func_obj.__name__.startswith("_"):
-        return f"{RETURN_VALUE} = " + f"tp.{func_obj.__name__}(**kwargs)"
-    elif _METHOD_OPS.get(func_obj.__name__, None):
+    if func_obj.__name__ in _METHOD_OPS.keys():
+        # Function needs to be executed in a specific way
         rtn_builder = _METHOD_OPS.get(func_obj.__name__, None)
-        args_list = list(kwargs.values())
-        if len(args_list) > 2:
-            print("WARNING WILL ONLY USE FIRST TWO PARAMS")
-        return f"{RETURN_VALUE} = " + rtn_builder(args_list[0], args_list[1])
+        api_call_locals[RETURN_VALUE] = rtn_builder(**kwargs)
     else:
-        raise RuntimeError(f"{fg('red')}Could not figure out method in function: {func_obj.__name__}{attr('reset')}")
+        # Execute API call normally.
+        exec(f"{RETURN_VALUE} = " + f"tp.{func_obj.__name__}(**kwargs)", globals(), api_call_locals)
+    # Print out debugging info.
+    print("API call: ", func_name, ", with parameters: ", kwargs)
+
 
+default_constraints_all = {"__rtruediv__": {"self": {"init": 1}},
```

Comment:
I don't understand why we need any keys besides `init`. e.g. we can replace `"shape": (3, 1)` with `tp.ones((3, 1))`. That also makes it less of a black box w.r.t what's actually being used by the test.

Reply:
The main issue that there are cases where we would also want to create a tensor such as `[3,2]` and if we have init assigned to use `tp.ones` then there would be no way to create tensor `[3,2]` without adding another keyword.

Reply:
I think I might have a solution for this. I'm just going to have another obj builder specifically for shape tensors.

Reply:
You can have:
```py
"func_name": {"self": tp.ones((3, 2))}
```
or:
```py
"func_name": {"self": tp.Tensor([3, 2])}
```
"init" can subsume all other arguments basically.

Reply:
And then I can just cast it to whatever dtype I need right?

Reply:
Yes